File: src/core/password/generator.js
const crypto = require('crypto');

/**
 * Service for generating cryptographically secure passwords and passphrases
 */
class PasswordGenerator {
  constructor() {
    // Character sets for different password types
    this.charsets = {
      uppercase: 'ABCDEFGHIJKLMNOPQRSTUVWXYZ',
      lowercase: 'abcdefghijklmnopqrstuvwxyz',
      numbers: '0123456789',
      symbols: '!@#$%^&*()_+-=[]{}|;:,.<>?'
    };
    
    // Words for passphrase generation (common but secure words)
    this.words = [
      'apple', 'beach', 'cloud', 'dream', 'earth', 'flame', 'grass', 'house', 
      'image', 'judge', 'knife', 'level', 'money', 'night', 'ocean', 'paper', 
      'queen', 'river', 'stone', 'table', 'unity', 'value', 'water', 'xenon', 
      'youth', 'zebra', 'album', 'brake', 'chair', 'dance', 'eagle', 'fault', 
      'globe', 'heart', 'inbox', 'jewel', 'kiosk', 'light', 'music', 'north',
      'opera', 'plant', 'quark', 'radio', 'sugar', 'tiger', 'urban', 'voice',
      'wheel', 'xylophone', 'yacht', 'zesty'
      // Full dictionary would be much larger in production
    ];
  }

  /**
   * Generate a cryptographically secure random password
   * @param {Object} options - Password generation options
   * @param {number} options.length - Password length (default: 16)
   * @param {boolean} options.uppercase - Include uppercase letters (default: true)
   * @param {boolean} options.lowercase - Include lowercase letters (default: true)
   * @param {boolean} options.numbers - Include numbers (default: true)
   * @param {boolean} options.symbols - Include symbols (default: true)
   * @param {boolean} options.excludeSimilar - Exclude similar characters like 1, l, I, 0, O (default: false)
   * @param {string} options.exclude - Characters to exclude (default: '')
   * @returns {string} - Generated password
   */
  generatePassword(options = {}) {
    // Default options
    const opts = {
      length: options.length || 16,
      uppercase: options.uppercase !== false,
      lowercase: options.lowercase !== false,
      numbers: options.numbers !== false,
      symbols: options.symbols !== false,
      excludeSimilar: options.excludeSimilar || false,
      exclude: options.exclude || ''
    };

    // Characters to exclude
    const excludeChars = opts.exclude.split('');
    if (opts.excludeSimilar) {
      excludeChars.push(...'1lI0O'.split(''));
    }

    // Build charset based on options
    let charset = '';
    if (opts.uppercase) {
      charset += this.charsets.uppercase;
    }
    if (opts.lowercase) {
      charset += this.charsets.lowercase;
    }
    if (opts.numbers) {
      charset += this.charsets.numbers;
    }
    if (opts.symbols) {
      charset += this.charsets.symbols;
    }

    // Remove excluded characters from charset
    charset = charset
      .split('')
      .filter(char => !excludeChars.includes(char))
      .join('');

    if (charset.length === 0) {
      throw new Error('No characters available for password generation');
    }

    // Generate password
    return this.generateRandomString(charset, opts.length);
  }

  /**
   * Generate a passphrase with random words
   * @param {Object} options - Passphrase generation options
   * @param {number} options.wordCount - Number of words (default: 4)
   * @param {string} options.separator - Word separator (default: '-')
   * @param {boolean} options.capitalize - Capitalize each word (default: false)
   * @returns {string} - Generated passphrase
   */
  generatePassphrase(options = {}) {
    // Default options
    const opts = {
      wordCount: options.wordCount || 4,
      separator: options.separator || '-',
      capitalize: options.capitalize || false,
      includeNumber: options.includeNumber || false,
      includeSymbol: options.includeSymbol || false
    };

    // Select random words
    const selectedWords = [];
    for (let i = 0; i < opts.wordCount; i++) {
      const randomIndex = this.getSecureRandomInt(0, this.words.length - 1);
      let word = this.words[randomIndex];
      
      if (opts.capitalize) {
        word = word.charAt(0).toUpperCase() + word.slice(1);
      }
      
      selectedWords.push(word);
    }

    // Build passphrase
    let passphrase = selectedWords.join(opts.separator);

    // Add a number if requested
    if (opts.includeNumber) {
      passphrase += opts.separator + this.getSecureRandomInt(0, 999);
    }

    // Add a symbol if requested
    if (opts.includeSymbol) {
      const symbols = this.charsets.symbols;
      const randomSymbol = symbols.charAt(this.getSecureRandomInt(0, symbols.length - 1));
      passphrase += randomSymbol;
    }

    return passphrase;
  }

  /**
   * Generate a cryptographically secure random string
   * @param {string} charset - Character set to use
   * @param {number} length - Length of the string
   * @returns {string} - Random string
   */
  generateRandomString(charset, length) {
    const randomChars = [];
    const charsetLength = charset.length;

    // Get secure random values
    const randomValues = crypto.randomBytes(length);

    // Convert to characters from the charset
    for (let i = 0; i < length; i++) {
      const randomIndex = randomValues[i] % charsetLength;
      randomChars.push(charset.charAt(randomIndex));
    }

    return randomChars.join('');
  }

  /**
   * Generate a cryptographically secure random integer
   * @param {number} min - Minimum value (inclusive)
   * @param {number} max - Maximum value (inclusive)
   * @returns {number} - Random integer
   */
  getSecureRandomInt(min, max) {
    const range = max - min + 1;
    
    // Ensure unbiased random number
    const bitsNeeded = Math.ceil(Math.log2(range));
    const bytesNeeded = Math.ceil(bitsNeeded / 8);
    const mask = Math.pow(2, bitsNeeded) - 1;
    
    let randomValue;
    do {
      const randomBytes = crypto.randomBytes(bytesNeeded);
      randomValue = 0;
      
      // Convert bytes to number
      for (let i = 0; i < bytesNeeded; i++) {
        randomValue = (randomValue << 8) | randomBytes[i];
      }
      
      // Apply mask to get only the bits we need
      randomValue = randomValue & mask;
    } while (randomValue >= range);
    
    return min + randomValue;
  }
}

module.exports = new PasswordGenerator(); 

File: src/password/generator.py
"""
Password generator.
Generates cryptographically secure passwords and passphrases.
"""

import os
import random
import string
import sys

# Add the project root to the Python path
sys.path.append(os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))))


class PasswordGenerator:
    """Generates secure random passwords and passphrases"""
    
    def __init__(self):
        """Initialize the password generator"""
        # Define character sets
        self.uppercase_chars = string.ascii_uppercase
        self.lowercase_chars = string.ascii_lowercase
        self.digit_chars = string.digits
        self.symbol_chars = "!@#$%^&*()-_=+[]{}|;:,.<>?/~"
        self.similar_chars = "iIl1Lo0O"
        
        # Load common words for passphrases
        self.common_words = [
            # Common English words for passphrases (4-8 letters)
            "apple", "baker", "cable", "dance", "eagle", "fable", "garden", "house",
            "igloo", "jumbo", "kiosk", "lemon", "music", "north", "ocean", "paper",
            "queen", "river", "sugar", "table", "uncle", "voice", "water", "xenon",
            "yacht", "zebra", "actor", "brave", "cloud", "dream", "event", "focus",
            "glass", "happy", "input", "judge", "knife", "light", "money", "night",
            "opera", "plane", "quiet", "radio", "storm", "train", "unity", "video",
            "woman", "youth", "album", "beach", "candy", "diary", "earth", "field",
            "guide", "horse", "image", "jewel", "kings", "lunch", "magic", "novel",
            "orbit", "peace", "quest", "robot", "space", "tiger", "urban", "value",
            "wagon", "xylophone", "young", "zesty", "amber", "basic", "coral", "draft",
            "enjoy", "flour", "grape", "hotel", "ivory", "joint", "kayak", "liver",
            "major", "noble", "oasis", "pearl", "quick", "royal", "solid", "trust",
            "ultra", "vital", "world", "xerox", "yield", "zones", "award", "blend",
            "craft", "dwell", "essay", "flute", "glide", "haste", "index", "joker",
            "knack", "lotus", "maker", "nerve", "onion", "pride", "quote", "relay"
        ]
    
    def generate_password(self, options=None):
        """
        Generate a secure random password
        
        Args:
            options (dict, optional): Options for password generation
                - length (int): Length of the password (default: 16)
                - uppercase (bool): Include uppercase letters (default: True)
                - lowercase (bool): Include lowercase letters (default: True)
                - numbers (bool): Include numbers (default: True)
                - symbols (bool): Include symbols (default: True)
                - exclude_similar (bool): Exclude similar characters (default: False)
                - exclude (str): Characters to exclude
                
        Returns:
            str: Generated password
        """
        # Set default options
        if options is None:
            options = {}
        
        length = options.get('length', 16)
        include_uppercase = options.get('uppercase', True)
        include_lowercase = options.get('lowercase', True)
        include_numbers = options.get('numbers', True)
        include_symbols = options.get('symbols', True)
        exclude_similar = options.get('exclude_similar', False)
        exclude_chars = options.get('exclude', '')
        
        # Prepare character pool
        char_pool = ""
        required_chars = []
        
        if include_uppercase:
            pool = self._remove_excluded_chars(self.uppercase_chars, exclude_chars, exclude_similar)
            char_pool += pool
            if pool:  # Only add a required character if there are characters left after exclusion
                required_chars.append(self._secure_random_choice(pool))
        
        if include_lowercase:
            pool = self._remove_excluded_chars(self.lowercase_chars, exclude_chars, exclude_similar)
            char_pool += pool
            if pool:
                required_chars.append(self._secure_random_choice(pool))
        
        if include_numbers:
            pool = self._remove_excluded_chars(self.digit_chars, exclude_chars, exclude_similar)
            char_pool += pool
            if pool:
                required_chars.append(self._secure_random_choice(pool))
        
        if include_symbols:
            pool = self._remove_excluded_chars(self.symbol_chars, exclude_chars, exclude_similar)
            char_pool += pool
            if pool:
                required_chars.append(self._secure_random_choice(pool))
        
        # Ensure we have characters to work with
        if not char_pool:
            raise ValueError("No characters available after applying exclusions")
        
        # Generate the password
        password = []
        
        # Add required characters first
        password.extend(required_chars)
        
        # Fill the rest with random characters
        for _ in range(length - len(required_chars)):
            password.append(self._secure_random_choice(char_pool))
        
        # Shuffle the password to mix the required characters randomly
        self._secure_shuffle(password)
        
        return ''.join(password)
    
    def generate_passphrase(self, options=None):
        """
        Generate a secure random passphrase
        
        Args:
            options (dict, optional): Options for passphrase generation
                - word_count (int): Number of words (default: 4)
                - capitalize (bool): Capitalize words (default: True)
                - include_number (bool): Include a number (default: True)
                - include_symbol (bool): Include a symbol (default: True)
                - delimiter (str): Word delimiter (default: '-')
                
        Returns:
            str: Generated passphrase
        """
        # Set default options
        if options is None:
            options = {}
        
        word_count = options.get('word_count', 4)
        capitalize = options.get('capitalize', True)
        include_number = options.get('include_number', True)
        include_symbol = options.get('include_symbol', True)
        delimiter = options.get('delimiter', '-')
        
        # Select random words
        words = []
        for _ in range(word_count):
            word = self._secure_random_choice(self.common_words)
            
            # Capitalize if enabled
            if capitalize:
                word = word.capitalize()
            
            words.append(word)
        
        # Add a number if enabled
        if include_number:
            position = self._secure_random_int(0, len(words))
            number = str(self._secure_random_int(0, 9999))
            words.insert(position, number)
        
        # Add a symbol if enabled
        if include_symbol:
            position = self._secure_random_int(0, len(words))
            symbol = self._secure_random_choice(self.symbol_chars)
            words.insert(position, symbol)
        
        # Join words with delimiter
        return delimiter.join(words)
    
    def _remove_excluded_chars(self, char_set, exclude_chars, exclude_similar):
        """
        Remove excluded characters from a character set
        
        Args:
            char_set (str): Character set to filter
            exclude_chars (str): Characters to exclude
            exclude_similar (bool): Whether to exclude similar characters
            
        Returns:
            str: Filtered character set
        """
        result = char_set
        
        # Remove explicitly excluded characters
        for char in exclude_chars:
            result = result.replace(char, '')
        
        # Remove similar characters if requested
        if exclude_similar:
            for char in self.similar_chars:
                result = result.replace(char, '')
        
        return result
    
    def _secure_random_choice(self, sequence):
        """
        Make a cryptographically secure random choice from a sequence
        
        Args:
            sequence (sequence): Sequence to choose from
            
        Returns:
            Any: Random element from the sequence
        """
        if not sequence:
            raise ValueError("Cannot choose from an empty sequence")
        
        return sequence[self._secure_random_int(0, len(sequence) - 1)]
    
    def _secure_random_int(self, min_value, max_value):
        """
        Generate a cryptographically secure random integer
        
        Args:
            min_value (int): Minimum value inclusive
            max_value (int): Maximum value inclusive
            
        Returns:
            int: Random integer
        """
        return random.SystemRandom().randint(min_value, max_value)
    
    def _secure_shuffle(self, sequence):
        """
        Shuffle a sequence using cryptographically secure randomness
        
        Args:
            sequence (list): Sequence to shuffle
            
        Returns:
            None: Shuffles in place
        """
        system_random = random.SystemRandom()
        system_random.shuffle(sequence)


# Singleton instance
password_generator = PasswordGenerator()


# Top-level functions for easier importing
def generate_password(options=None):
    """
    Generate a secure random password (convenience function)
    
    Args:
        options (dict, optional): Options for password generation
            
    Returns:
        str: Generated password
    """
    return password_generator.generate_password(options)


def generate_passphrase(options=None):
    """
    Generate a secure random passphrase (convenience function)
    
    Args:
        options (dict, optional): Options for passphrase generation
            
    Returns:
        str: Generated passphrase
    """
    return password_generator.generate_passphrase(options) 